Update PREREQUISITE.md (#185)
euphony-io/euphony: PREREQUISITE.md

dependencies {
	// other dependencies
	// ...

> 	implementation 'co.euphony.lib:euphony:0.8.0.1'

Reviewer: LGTM 👍

> OS : Windows 10  
> Android Studio : Arctic Fox | 2020.3.1

1. Download `euphony.aar` : [MavenCentral euphony artifact](https://search.maven.org/artifact/co.euphony.lib/euphony/0.8.0.1/aar) follow the link and download aar file

<img width="392" alt="aar_001" src="https://user-images.githubusercontent.com/27720475/130188260-a0f3ee6d-7afc-4dfc-928e-5562eca476ba.png">

2. Put `euphony.aar` file in `libs` folder, and just click `Refactor`

<img width="392" alt="aar_002_auto_x2_colored_toned" src="https://user-images.githubusercontent.com/27720475/130187177-b97b55ef-158a-4975-b0f8-b9e8bfdc5886.png">

> <img width="392" alt="aar_003_auto_x2_colored_toned" src="https://user-images.githubusercontent.com/76645322/184471738-64d3dd96-73e2-4de8-9dc6-133100e5470d.png">

Author: I changed image for matching version.

Let's look up the code.

```java

> package com.exmaple.euphonytest;

Reviewer: ```suggestion
package com.example.euphonytest;
```

Here is some typo :)

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {
    private ActivityResultLauncher requestPermissionLauncher;
    private Button button;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        requestPermissionLauncher = registerForActivityResult(
                new ActivityResultContracts.RequestPermission(),
                new ActivityResultCallback<Boolean>() {
                    @Override
                    public void onActivityResult(Boolean result) {
                        if (result) {
                            // PERMISSION GRANTED
                        } else {
                            // PERMISSION DENIED
                        }
                    }
                });

> 

Author: I use ActivityResultLauncher instead of requestPermissions.
I'm not sure this is best way, so please tell me if there are other good ways.

        button = findViewById(R.id.button2);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    final String permission = Manifest.permission.RECORD_AUDIO;

>                     requestPermissionLauncher.launch(permission);

Author: I use ActivityResultLauncher instead of requestPermissions.
I'm not sure this is best way, so please tell me if there are other good ways.

                    if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        requestPermissions(new String[]{
                                Manifest.permission.RECORD_AUDIO}, 1000);
                    }
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    final String permission = Manifest.permission.RECORD_AUDIO;
                    requestPermissionLauncher.launch(permission);
                }
            }
        });
    }
```

```java
   //after requestPermission
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if(requestCode == 1000) {
            boolean check_result = true;
            for (int result : grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    check_result = false;
                    break;
                }
            }
            if(check_result == true) {

            } else {
                finish();
            }
        }

>     }

Reviewer: Thanks for removal of deprecated function 👍

>     requestPermissionLauncher.launch(permission);

Author: I use ActivityResultLauncher instead of requestPermissions.
I'm not sure this is best way, so please tell me if there are other good ways.

>     requestPermissionLauncher.launch(permission);

Author: I use ActivityResultLauncher instead of requestPermissions.
I'm not sure this is best way, so please tell me if there are other good ways.

Reviewer: Thanks for replacing it with the latest Android API :) 
Single permission is fine, but the need for multiplePermissions seems to be high. 
How about adding a usage for multiplePermissions as well?!

```java
package com.example.euphonytest;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {
    private Button button;

>     private ActivityResultLauncher<String[]> multiplePermissionLauncher;
>     final String[] permissions = {
>             Manifest.permission.RECORD_AUDIO
>             // add other permissions
>     };
> 
>     @Override
>     protected void onCreate(Bundle savedInstanceState) {
>         super.onCreate(savedInstanceState);
>         setContentView(R.layout.activity_main);
> 
>         registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), isGranted -> {
>             if(isGranted.toString().contains("false")) {
>                 Log.d("Permissions", "result: " + isGranted.toString());
>                 finish();
>             }
>         });

Reviewer 2: ```suggestion
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        String[] permissions = {
                Manifest.permission.RECORD_AUDIO
                // add other permissions
        };

        // after requestPermissions
        ActivityResultLauncher<String[]> multiplePermissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), grantResults -> {
            for(Boolean result : grantResults.values()) {
                if (!result) {
                    finish();
                }
            }
        });
```

        button = findViewById(R.id.button2);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    multiplePermissionLauncher.launch(permissions);
                }
            }
        });
    }

}
```

This is `MainActivity` for adding permission.

```java

> ActivityResultLauncher<String[]> multiplePermissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), isGranted -> {
>     if(isGranted.toString().contains("false")) {
>         Log.d("Permissions", "result: " + isGranted.toString());
>         finish();
>     }
> });
> 
> if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
>     multiplePermissionLauncher.launch(permissions);
> }

Reviewer 2: ```suggestion
String[] permissions = {
    Manifest.permission.RECORD_AUDIO
    // add other permissions
};

// after requestPermissions
ActivityResultLauncher<String[]> multiplePermissionLauncher = registerForActivityResult(new 
ActivityResultContracts.RequestMultiplePermissions(), grantResults -> {
    for(Boolean result : grantResults.values()) {
        if (!result) {
            finish();
        }
    }
});
if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
    multiplePermissionLauncher.launch(permissions);
}
```

Author: LGTM !

I understand where I was making a mistake, It seems to be more readable than the example I wrote. 
I really appreciate your help ! 😄

```

> We can know whether permission is denied or not. And we also consider both sdk version is higher than 16 which is mashmellow and call function requestPermission.

Reviewer 2: ```suggestion
We can know whether permissions are denied or not. And we also consider both sdk version is higher than 16 which is marshmallow and call function RequestMultiplePermissions.
```

Author: LGTM 👍
